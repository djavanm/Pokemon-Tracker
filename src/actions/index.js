export const setPokemon = pokemon => ({
  type: 'SET_POKEMON',
  pokemon
});

export const catchPokemon = pokemon => ({
  type: 'CATCH_POKEMON',
  pokemon
});

export const releasePokemon = id => ({
  type: 'RELEASE_POKEMON',
  id
});
