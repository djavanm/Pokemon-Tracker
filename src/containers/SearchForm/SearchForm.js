import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { catchPokemon } from '../../actions';
import { fetchPokemon } from '../../thunks/fetchPokemon';
import { randNum } from '../../helpers';

export class SearchForm extends Component {
  constructor() {
    super()
    this.state = {
      name: '',
      id: '',
      newPokemon: false
    }
  };

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  handleSubmit = async e => {
    e.preventDefault();
    const { fetchPokemon } = this.props;
    const { id } = this.state;
    const url = `https://pokeapi.co/api/v2/pokemon/${id}/`;
    fetchPokemon(url);
    this.clearInputs();
  };

  handleSurprise = async e => {
    e.preventDefault();
    const { fetchPokemon } = this.props;
    const newId = randNum();
    const url = `https://pokeapi.co/api/v2/pokemon/${newId}/`;
    fetchPokemon(url);
    this.clearInputs();
  };

  clearInputs = () => {
    this.setState({ name: '', id: '', newPokemon: true });
  };

  showNew = async (e) => {
    e.preventDefault();
    const { fetchPokemon, pokemon } = this.props;
    const id = pokemon.id;
    const newId = e.target.name === 'next' ? id + 1 : id - 1;
    const url = `https://pokeapi.co/api/v2/pokemon/${newId}/`;
    fetchPokemon(url);
  };

  handleCatchPokemon = (e) => {
    e.preventDefault();
    const { pokemon, catchPokemon } = this.props;
    catchPokemon(pokemon);
  };


  render() {
    let displayBool = true;
    let catchBtnText = '';
    const { name, id, newPokemon } = this.state;
    const { pokemon, caughtPokemon } = this.props;
    const placeholderName = newPokemon ? pokemon.name : 'Enter a Name!';
    const placeholderId = newPokemon ? `#${pokemon.id}` : 'Enter a Number!';
    if(pokemon) {
      displayBool = caughtPokemon.map(pokemon => pokemon.id).includes(pokemon.id);
      catchBtnText =  displayBool ? 'Caught!' : 'Catch!';
    };
    return (
      <form className='search-form'>
        <button className="scroll-button" onClick={this.showNew} name="previous">{`<`}</button>
        <div className = 'middle-container'>
        <div className='inputs-container'>
          <input className="search-name input" type="text" placeholder={placeholderName} value={name} onChange={this.handleChange} name="name"/>
          <input className="search-num input" type="number" placeholder={placeholderId} value={id} onChange={this.handleChange} name="id" />
        </div>
        <div className="button-container">
          <button className="search-btn" type="submit" onClick={this.handleSubmit}> Search </button>
          <button className="search-btn" type="submit" onClick={this.handleSurprise}> Surprise Me! </button>
          {pokemon && <button className="search-btn" disabled={displayBool} onClick={this.handleCatchPokemon}> {catchBtnText}</button>}
        </div>
        </div>
        <button className="scroll-button" onClick={(e) => this.showNew(e)} name="next" >{`>`}</button>
      </form>
    )
  }
};

export const mapStateToProps = state => ({
  pokemon: state.currentPokemon,
  caughtPokemon: state.caughtPokemon
});

export const mapDispatchToProps = dispatch => (
  bindActionCreators({ fetchPokemon, catchPokemon }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(SearchForm);
