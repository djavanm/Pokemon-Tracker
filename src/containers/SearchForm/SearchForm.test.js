import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm, mapStateToProps, mapDispatchToProps } from './SearchForm';
import { setPokemon } from '../../actions';

describe('SearchForm', () => {
  let wrapper;
  const mockPokemon = {
    name: 'Bulbasaur',
    description: 'Bulbasaur is cool.',
    gen: 'Generation 1',
    gif: 'https://projectpokemon.org/images/normal-sprite/bulbasaur.gif',
    height: 7,
    id: 1,
    image: 'https://assets.pokemon.com/assets/cms2/img/pokedex/full/001.png',
    sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png'
  };

  const mockCaughtPokemon = [mockPokemon];
  const mockTeamPokemon = [mockPokemon];
  const mockCurrentPokemon =  mockPokemon;
  const mockState = {
    currentPokemon: mockPokemon,
    caughtPokemon: mockCaughtPokemon,
    teamPokemon: mockTeamPokemon
  };

  beforeEach(() => {
    wrapper = shallow(<SearchForm
      pokemon={mockCurrentPokemon}
      caughtPokemon={mockCaughtPokemon}
      />)
  });

  it('should match the snapshot given the proper information', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update the state when handle change is called', () => {
    const mockEventName = {
      target: {
        name: 'name',
        value: 'squirtle'
      }
    };
    const mockEventId= {
      target: {
        name: 'id',
        value: 4
      }
    };
    const expectedName = 'squirtle';
    const expectedId = 4;
    wrapper.instance().handleChange(mockEventName);
    wrapper.instance().handleChange(mockEventId);
    expect(wrapper.state('name')).toEqual(expectedName);
    expect(wrapper.state('id')).toEqual(expectedId);
  });

  it('mapStateToProps should grab the props it needs', () => {
    const expected = {
      pokemon: mockPokemon,
      caughtPokemon: mockCaughtPokemon
    };
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });

  it('it calls dispatch with the setPokemon action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setPokemon(mockPokemon);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setPokemon(mockPokemon);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
