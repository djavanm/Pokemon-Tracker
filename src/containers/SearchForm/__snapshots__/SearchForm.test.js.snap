// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot given the proper information 1`] = `
<form
  className="search-form"
>
  <button
    className="scroll-button"
    name="previous"
    onClick={[Function]}
  >
    &lt;
  </button>
  <div
    className="middle-container"
  >
    <div
      className="inputs-container"
    >
      <input
        className="search-name input"
        name="name"
        onChange={[Function]}
        placeholder="Enter a Name!"
        type="text"
        value=""
      />
      <input
        className="search-num input"
        name="id"
        onChange={[Function]}
        placeholder="Enter a Number!"
        type="number"
        value=""
      />
    </div>
    <div
      className="button-container"
    >
      <button
        className="search-btn"
        onClick={[Function]}
        type="submit"
      >
         Search 
      </button>
      <button
        className="search-btn"
        onClick={[Function]}
        type="submit"
      >
         Surprise Me! 
      </button>
      <button
        className="search-btn"
        disabled={true}
        onClick={[Function]}
      >
         
        Caught!
      </button>
    </div>
  </div>
  <button
    className="scroll-button"
    name="next"
    onClick={[Function]}
  >
    &gt;
  </button>
</form>
`;
